src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -52,80 +74,188 @@ private static SnowflakeDateTimeFormat createDateTimeFormatter() {
     }
+    if (o instanceof String || o instanceof Boolean || o instanceof Character) {
+      return true;
+    }
+    if (o instanceof Long
+        || o instanceof Integer
+        || o instanceof Short
+        || o instanceof Byte
+        || o instanceof BigInteger) {
+      return true;
+    }
+    if (o instanceof Float || o instanceof Double || o instanceof BigDecimal) {
+      return true;
+    }
+    if (o instanceof java.time.LocalTime || o instanceof OffsetTime) {
+      return true;
+    }
+    if (o instanceof LocalDate
+        || o instanceof LocalDateTime
+        || o instanceof ZonedDateTime
+        || o instanceof OffsetDateTime) {
+      return true;
+    }
+    if (o instanceof Map) {
+      boolean allKeysAreStrings =
+          ((Map<?, ?>) o).keySet().stream().allMatch(x -> x instanceof String);
+      if (!allKeysAreStrings) {
+        return false;
+      }
+      boolean allValuesAreAllowed =
+          ((Map<?, ?>) o).values().stream().allMatch(DataValidationUtil::isAllowedVariantType);
+      return allValuesAreAllowed;
+    }
+    if (o instanceof byte[]
+        || o instanceof short[]
+        || o instanceof int[]
+        || o instanceof long[]
+        || o instanceof float[]
+        || o instanceof double[]
+        || o instanceof boolean[]
+        || o instanceof char[]) {
+      return true;
+    }
+    if (o.getClass().isArray()) {
+      return Arrays.stream((Object[]) o).allMatch(DataValidationUtil::isAllowedVariantType);
+    }
+    if (o instanceof List) {
+      return ((List<?>) o).stream().allMatch(DataValidationUtil::isAllowedVariantType);
+    }
+    return false;
+  }
 
-    String output;
-    try {
-      output = objectMapper.writeValueAsString(input);
-    } catch (Exception e) {
-      throw new SFException(
-          e,
-          ErrorCode.INVALID_ROW,
-          input.toString(),
-          "Input column can't be convert to a valid string");
+  /**
+   * Validates and parses JSON array. Non-array types are converted into single-element arrays. All
+   * types in the array tree must be valid variant types, see {@link
+   * DataValidationUtil#isAllowedVariantType}.
+   *
+   * @param input Object to validate
+   * @return JSON array representing the input
+   */
+  static String validateAndParseArray(Object input) {
+    JsonNode jsonNode = validateAndParseVariantAsJsonTree(input, "ARRAY");
+
+    // Non-array values are ingested as single-element arrays, mimicking the Worksheets behavior
+    if (!jsonNode.isArray()) {
+      jsonNode = objectMapper.createArrayNode().add(jsonNode);
     }
 
+    String output = jsonNode.toString();
     // Throw an exception if the size is too large
-    if (output.length() > BYTES_16_MB) {
-      throw new SFException(
-          ErrorCode.INVALID_ROW,
-          input.toString(),
-          String.format("Array too large. length=%d maxLength=%d", output.length(), BYTES_16_MB));
+    int stringLength = output.getBytes(StandardCharsets.UTF_8).length;
```

Comment:
do you know if `StandardCharsets.UTF_8` is the only thing we support in SF?

Reply:
It is, according to the documentation: https://docs.snowflake.com/en/sql-reference/collation.html#collation-not-supported-for-strings-in-variant-array-or-object